service/history/ndc/hsm_state_replicator.go

```diff
@@ -0,0 +1,271 @@
+	}
+
+	lastItem, err := versionhistory.GetLastVersionHistoryItem(request.EventVersionHistory)
+	if err != nil {
+		return err
+	}
+
+	workflowContext, release, err := r.workflowCache.GetOrCreateWorkflowExecution(
+		ctx,
+		r.shardContext,
+		namespaceID,
+		execution,
+		workflow.LockPriorityHigh,
+	)
+	if err != nil {
+		return err
+	}
+	defer func() { release(retError) }()
+
+	mutableState, err := workflowContext.LoadMutableState(ctx, r.shardContext)
+	if err != nil {
+		if _, isNotFound := err.(*serviceerror.NotFound); isNotFound {
+			return serviceerrors.NewRetryReplication(
+				"sync HSM state encountered workflow not found",
+				namespaceID.String(),
+				execution.GetWorkflowId(),
+				execution.GetRunId(),
+				common.EmptyEventID,
+				common.EmptyVersion,
+				lastItem.EventId+1, // underlying resend logic is exclusive-exclusive
+				lastItem.Version,
+			)
+		}
+		return err
+	}
+
+	synced, err := r.syncHSMNode(mutableState, request)
+	if err != nil || !synced {
+		return err
+	}
+
+	state, _ := mutableState.GetWorkflowStateStatus()
+	if state == enumsspb.WORKFLOW_EXECUTION_STATE_COMPLETED {
+		return workflowContext.SubmitClosedWorkflowSnapshot(
+			ctx,
+			r.shardContext,
+			workflow.TransactionPolicyPassive,
+		)
+	}
+
+	updateMode := persistence.UpdateWorkflowModeUpdateCurrent
+	if state == enumsspb.WORKFLOW_EXECUTION_STATE_ZOMBIE {
+		updateMode = persistence.UpdateWorkflowModeBypassCurrent
+	}
+
+	return workflowContext.UpdateWorkflowExecutionWithNew(
+		ctx,
+		r.shardContext,
+		updateMode,
+		nil, // no new workflow
+		nil, // no new workflow
+		workflow.TransactionPolicyPassive,
+		nil,
+	)
+}
+
+func (r *HSMStateReplicatorImpl) syncHSMNode(
+	mutableState workflow.MutableState,
+	request *shard.SyncHSMRequest,
+) (bool, error) {
+
+	shouldSync, err := r.compareVersionHistory(mutableState, request.EventVersionHistory)
+	if err != nil || !shouldSync {
+		return shouldSync, err
+	}
+
+	currentHSM := mutableState.HSM()
+
+	// we don't care about the root here which is the entire mutable state
+	incomingHSM, err := hsm.NewRoot(
+		r.shardContext.StateMachineRegistry(),
+		workflow.StateMachineType.ID,
+		mutableState,
+		request.StateMachineNode.Children,
+		mutableState,
+	)
+
+	synced := false
+	incomingHSM.Walk(func(incomingNode *hsm.Node) error {
+
+		if incomingNode.Parent == nil {
+			// skip root which is the entire mutable state
+			return nil
+		}
+
+		incomingNodePath := incomingNode.Path()
+		currentNode, err := currentHSM.Child(incomingNodePath)
+		if err != nil {
+			// we already done history resend before, node should always be found
+			// and node creation always associated with an event right now
+			// TODO: support node creation without event
+			return err
+		}
+
+		if incomingNode.InternalRepr().InitialNamespaceFailoverVersion !=
+			currentNode.InternalRepr().InitialNamespaceFailoverVersion {
+			return nil
+		}
+
+		currentNodeSynced, err := currentNode.Sync(incomingNode)
```

Comment:
We should compare the last update namespace failover version as you suggested before as a tie breaker.
Otherwise, for callbacks and operation cancelation, one cluster may transition to failed state and the other to succeeded state and there's no way to implement a comparator between those two states.

Reply:
Syned offline. Will do in a separate PR